### DISCUSSION TITLE
New Row Violates Row-Level Security Policy Error with Presigned Upload URLs

### DISCUSSION THREAD

**Discussion by martinbartolo** [supabase/supabase#18947] [Questions]
I am trying to enable client-side uploads to a storage bucket using presigned upload URLs created by my server-side supabase client. However, I am running into an issue where using the signed URL for uploads using axios (since I would like to track upload progress and be able to abort uploads) gives me the following error

```json
{
    "statusCode": "403",
    "error": "Unauthorized",
    "message": "new row violates row-level security policy"
}
```

I am getting my pre-signed URLs as follows on the server side. This seems to work since I get back a signed URL to upload to.

```js
const supabase = createClient(env.SUPABASE_URL, env.SUPABASE_SERVICE_ROLE_KEY);

export const supabaseRouter = createTRPCRouter({
  getUploadURL: protectedProcedure
    .input(
      z.object({
        bucket: z.string().optional(),
        filename: z.string().optional(),
        userId: z.string().optional(),
      }),
    )
    .query(async ({ ctx, input }) => {
      if (!input.bucket || !input.filename || !input.userId) {
        return null;
      }
      if (ctx.session.user.id !== input.userId) {
        throw new Error("User is not authorized to upload to this bucket");
      }
      const { data, error } = await supabase.storage
        .from(input.bucket)
        .createSignedUploadUrl(`${input.userId}/${input.filename}`);
      if (error) {
        throw error;
      }
      return data;
    }),
});
```

Now when I try to use the url and the token from the presigned URL with this axios POST request client-side I get the 403 error.
```js
const uploadToURL = async (file: File, url: string, token: string) => {
      await axios
        .post(url, file, {
          headers: {
            "Content-Type": file.type,
            authorization: `Bearer ${token}`,
          },
          onUploadProgress: (progressEvent) => {
            if (progressEvent.total) {
              const progress =
                (progressEvent.loaded / progressEvent.total) * 100;
              setUploadProgress!(progress);
            }
          },
          signal: abortController.current?.signal,
        })
        .then((response) => console.log(response))
        .catch((error) => console.log(error));
    };
``` 
Since the [createSignedUploadUrl](https://supabase.com/docs/reference/javascript/storage-from-createsigneduploadurl) function returns a signed URL, a path and a token, I have tried using the signed URL without any authorization header, the signed URL with the token as authorization header and the path with the token as authorization header and I get the same error regardless.

My bucket has no policies assigned but [uploadToSignedUrl](https://supabase.com/docs/reference/javascript/storage-from-uploadtosignedurl) function does not require any RLS policies so I don't think that is an issue. 

I also have not initialized a client-side supabase client since I am using axios. so maybe this could be the problem? I would love to figure out what I am getting wrong here or if I am taking the wrong approach to uploading to a presigned URL.

**Comment by GaryAustin1** [comment]
EDIT redid entire post.
You should need to meet the insert policy to create the signedUrl.   After that you should be able to use it calling the correct endpoint without an authorization header.

The API docs don't seem to be working so can't verify anything else....
https://supabase.github.io/storage-api/

**Reply by martinbartolo** [reply]
On the server side I am using my service role secret key to create my supabase client so that bypasses the RLS and gives me back the signedURL. I am getting them back in the form https://<project_id>.supabase.co/storage/v1/object/upload/sign/<bucket>/<folder/filename>?token=< token123 > and then using this url in my axios request.

There is this docs page but I don't think it is up to date [https://supabase.com/docs/reference/storage/usage?response-status=200](https://supabase.com/docs/reference/storage/usage?response-status=200)

**Reply by GaryAustin1** [reply]
If your expiration time is for sure good then you might want to generate an issue in supabase/storage-api.
The owner is used as part of the token and RLS is checked when you create the upload url.
I wonder if there is a bug if owner is null, which it would be for service_role...
Not sure how many users generate upload url with service_role versus as a signed in user.

**Reply by martinbartolo** [reply]
Will do, thanks for your replies. It definitely isn't an expiration time issue since [createSignedUploadUrl](https://supabase.com/docs/reference/javascript/storage-from-createsigneduploadurl) says that URLs take 2 hours to expire and I am trying to use them right after they are generated. The reason that I am using service_role is because I'm using nextauth instead of supabase auth so having my server generate presigned urls at an endpoint only accessible to logged-in users seems like the best way to interact with my storage buckets.
